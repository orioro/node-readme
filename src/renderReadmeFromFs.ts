import * as path from 'path'
import { writeFile } from 'fs'
import { promisify } from 'util'
import { Transform } from 'readable-stream'
import vinylFs from 'vinyl-fs'
import rename from 'gulp-rename'
import { parseCommentsFromFs } from './parseComment'
import toc from 'markdown-toc'
import { uniq } from 'lodash'
import { templateRender } from './template'
import { renderCommentTitle } from './renderComment'

const writeFileAsync = promisify(writeFile)

const REFERRABLE_COMMENT_TYPES = [
  'function', 'func', 'method', 'callback',
  'class', 'constructor',
  'constant', 'const',
  'event',
  'global',
  'module',
  'typedef'
]

const strReplaceAll = (str, search, replacement) => {
  return str.replace(new RegExp(search, 'g'), replacement)
}

const linkInnerTypeReferences = (comments, text) => {
  return comments.reduce((acc, comment) => (
    REFERRABLE_COMMENT_TYPES.includes(comment.commentType)
      ? acc.replace(
          new RegExp(`{.*?${comment.name}.*?}`, 'g'),
          match => match.replace(
            comment.name,
            `[${comment.name}](#${toc.slugify(renderCommentTitle(comment))})`
          )
        )
      : acc
  ), text) 
}

export const vinylReadmeRenderStream = (context) => {
  return new Transform({
    objectMode: true,
    transform(file, enc, cb) {
      if (file.isStream()) {
        this.emit('error', Error('@orioro/readme: Streams files are not supported'))
        return cb()
      } else if (file.isBuffer()) {
        const rendered = linkInnerTypeReferences(
          context.comments,
          templateRender(
            file.contents.toString('utf8'),
            context
          )
        )

        file.contents = Buffer.from(
          rendered + '\n',
          'utf8'
        )

        this.push(file)
      }
      
      cb()
    }
  })
}

const readMeRenderContext = comments => {
  const files = uniq(comments.map(comment => comment.file))
  const todo = comments.reduce((acc, comment) => {
    return Array.isArray(comment.todo) && comment.todo.length > 0
      ? [...acc, ...comment.todo.map(todo => ({
          ...todo,
          source: comment.file
        }))]
      : acc
  }, [])

  return {
    comments,
    todo,
    files: files.reduce((acc, file) => {
      const _comments = comments
        .filter(comment => comment.file === file)

      const moduleComment = _comments.find(comment => comment.commentType === 'module')

      return {
        ...acc,
        [file]: {
          name: moduleComment ? moduleComment.name : file,
          comments: _comments.reduce((acc, comment, index) => {
            const name = comment.name || index + ''
            return {
              ...acc,
              [name]: comment
            }
          }, {})
        }
      }
    }, {})
  }
}

const renderTodo = (targetPath, todo) => {
  const longestNameCharCount = todo.reduce((acc, todo) => (
    todo.name.length > acc
      ? todo.name.length
      : acc
  ), 0)

  return writeFileAsync(
    targetPath,
    todo
      .map(todo => {
        const name = (todo.name || '').trim().padEnd(longestNameCharCount, ' ')
        const description = (todo.description || '').trim()
        const source = todo.source.trim()

        return `- ${name} | ${description} (${source})`
      })
      .sort()
      .join('\n') + '\n'
  )
}

/**
 * Generates the readme.md file
 *
 * @function renderReadmeFromFs
 * @param {Object} config
 * @param {GlobPattern[]} config.templatesSrc
 * @param {GlobPattern[]} config.commentsSrc
 * @param {string} config.cwd
 * @param {string} [config.dest='./']
 * @return {Promise<void>}
 */
export const renderReadmeFromFs = ({
  templatesSrc,
  commentsSrc,
  dest = './',
  todo = true,
  cwd
}) => (
  parseCommentsFromFs(commentsSrc, { cwd })
  .then(comments => {

    const context = readMeRenderContext(comments)

    const readmePromise = new Promise((resolve, reject) => {
      const stream = vinylFs
        .src(templatesSrc, {
          cwd,
          dot: true
        })
        .pipe(vinylReadmeRenderStream(context))
        .pipe(rename(path => ({
          ...path,
          basename: path.basename.replace(/^\./, ''),
        })))
        .pipe(vinylFs.dest(dest, { cwd }))

      stream.on('finish', resolve)
      stream.on('error', reject)
    })

    const todoPromise = todo
      ? renderTodo(
          path.join(cwd, dest, 'TODO.md'),
          context.todo
        )
      : Promise.resolve()

    return Promise.all([readmePromise, todoPromise]).then(() => undefined)
  })
)
