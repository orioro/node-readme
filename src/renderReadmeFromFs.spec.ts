import * as fs from 'fs'
import * as path from 'path'
import { renderReadmeFromFs } from './renderReadmeFromFs'

const FIXTURES_PATH = path.join(__dirname, '../test')
const TMP_PATH = path.join(__dirname, '../tmp')

describe('renderReadmeFromFs', () => {
  test('', () => {
    return renderReadmeFromFs({
      templatesSrc: ['**/.*.md'],
      commentsSrc: [
        'src/**/*.js',
        'src/**/*.jsx',
        'src/**/*.ts',
        'src/**/*.tsx',
      ],
      cwd: path.join(FIXTURES_PATH, 'fixture-1'),
    })
    .then(() => {
      expect(fs.readFileSync(path.join(FIXTURES_PATH, 'fixture-1/README.md'), 'utf8'))
        .toMatchSnapshot()
      expect(fs.readFileSync(path.join(FIXTURES_PATH, 'fixture-1/TODO.md'), 'utf8'))
        .toMatchSnapshot()
    })
  })
})
