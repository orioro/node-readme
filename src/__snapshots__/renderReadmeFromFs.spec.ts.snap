// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renderReadmeFromFs  1`] = `
"# readme

Cras pharetra massa ac est convallis, id placerat nulla auctor. Pellentesque fermentum, lectus ut porttitor egestas, ligula mi rhoncus est, sit amet aliquet purus dolor eget nisi. Nunc sodales lectus augue, id porta ligula euismod non. In volutpat pellentesque purus bibendum porttitor.

# API

- src/index.ts
  - [\`CustomType\`](#customtype)
  - [\`func1(strParam, numParam)\`](#func1strparam-numparam)
  - [\`func2(param1, param2)\`](#func2param1-param2)
- src/someMethod.ts
  - [\`someMethod(param)\`](#somemethodparam)


##### \`CustomType\`

Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris
nisi ut aliquip ex ea commodo consequat.

- \`name\` {String}
- \`bool\` {Boolean}

##### \`func1(strParam, numParam)\`

Lorem ipsum dolor sit amet, consectetur adipiscing elit,
sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.

Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris
nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in
reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla
pariatur.

\`\`\`js
const result = func1('some string', 10)
// false
\`\`\`

Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.

- \`strParam\` {string}
- \`numParam\` {number}
- Returns: \`result\` {boolean} Nam libero tempore, cum soluta nobis est
                          eligendi optio.

##### \`func2(param1, param2)\`

Aliquam at metus molestie, luctus arcu id, ornare ligula.
View \`func1\`

- \`param1\` {Object}
  - \`opt1\` {[CustomType](#customtype)}
  - \`opt2\` {string}
- \`param2\` {[CustomType](#customtype)}
- Returns: \`result\` {[CustomType](#customtype) | null} 

##### \`someMethod(param)\`

- \`param\` {number}
- Returns: \`result\` {number} Lorem ipsum dolor sit amet, consectetur adipiscing
"
`;

exports[`renderReadmeFromFs  2`] = `
"- #3     | Some issue (src/index.ts)
- -      | Something to be done (src/someMethod.ts)
- #45:OK | Something to be done that already has an issue number (src/someMethod.ts)
"
`;
